pkg/splunk/enterprise/util.go

```diff
@@ -292,3 +294,236 @@ func DeleteOwnerReferencesForS3SecretObjects(client splcommon.ControllerClient,
+			appDeployContext.IsDeploymentInProgress = true
+		}
+
+		// Finally update the Map entry with latest info
+		appDeployContext.AppsSrcDeployStatus[appSrc] = appSrcDeploymentInfo
+	}
+
+	return true, err
+}
+
+// isAppExtentionValid checks if an app extention is supported or not
+func isAppExtentionValid(receivedKey string) bool {
+	appExtIdx := strings.LastIndex(receivedKey, ".")
+	if appExtIdx < 0 {
+		return false
+	}
+
+	switch appExt := receivedKey[appExtIdx+1:]; appExt {
+	case "spl":
+		return true
+
+	case "tgz":
+		return true
+
+	default:
+		return false
+	}
+}
+
+// createOrUpdateAppSrcDeploymentInfo  modifies the App deployment status as perceived from the remote object listing
+func createOrUpdateAppSrcDeploymentInfo(appSrcDeploymentInfo *enterprisev1.AppSrcDeployInfo, remoteS3ObjList []*splclient.RemoteObject) bool {
+	scopedLog := log.WithName("createOrUpdateAppSrcDeploymentInfo").WithValues("Called with length: ", len(remoteS3ObjList))
+
+	var found bool
+	var appName string
+	var newAppInfoList []enterprisev1.AppDeploymentInfo
+	var appChangesDetected bool
+	var appDeployInfo enterprisev1.AppDeploymentInfo
+
+	for _, remoteObj := range remoteS3ObjList {
+		receivedKey := *remoteObj.Key
+		if !isAppExtentionValid(receivedKey) {
+			scopedLog.Info("App name Parsig", " Ignoring the key: %s, with invalid extention for the app", receivedKey)
+			continue
+		}
+
+		nameAt := strings.LastIndex(receivedKey, "/")
+		if nameAt == -1 {
+			appName = receivedKey
+		} else {
+			appName = receivedKey[nameAt+1:]
+		}
+
+		// Now update App status as seen in the remote listing
+		found = false
+		for _, appDeployInfo = range appSrcDeploymentInfo.AppDeploymentInfoList {
+			if appDeployInfo.AppName == appName {
+				found = true
+				if appDeployInfo.ObjectHash != *remoteObj.Etag {
+					appDeployInfo.ObjectHash = *remoteObj.Etag
+					appDeployInfo.DeployStatus = enterprisev1.DeployStatusPending
+					appChangesDetected = true
+				} else if appDeployInfo.RepoState == enterprisev1.RepoStateDeleted {
+					appDeployInfo.ObjectHash = *remoteObj.Etag
+					appDeployInfo.RepoState = enterprisev1.RepoStateActive
+					appDeployInfo.DeployStatus = enterprisev1.DeployStatusPending
+					appChangesDetected = true
+				}
+
+				// Found the App and finished the needed work. we can break here
+				break
+			}
+		}
+
+		// Update our local list if it is a new app
+		if !found {
+			appDeployInfo.AppName = appName
+			appDeployInfo.ObjectHash = *remoteObj.Etag
+			appDeployInfo.Size = uint64(*remoteObj.Size)
+			appDeployInfo.RepoState = enterprisev1.RepoStateActive
+			appDeployInfo.DeployStatus = enterprisev1.DeployStatusPending
+			appChangesDetected = true
+
+			// Add it to a seperate list so that we don't loop through the newly added entries
+			newAppInfoList = append(newAppInfoList, appDeployInfo)
+		}
+	}
+
+	// Add the newly discovered Apps to the App source group
```

Comment:
May be do this only when `newAppInfoList` is not empty?

Reply:
Go auto covers in case of nil list. i.e. the existing list is not re-allocated as long as the existing length doesn't cross its capacity.